src/scriptrag/agents/context_query.py

```diff
@@ -0,0 +1,439 @@
+                                params.next_scene_hash = next_scene.content_hash
+                        break
+
+        # Extract project name from settings or file path
+        if settings:
+            params.project_name = getattr(settings, "project_name", None)
+
+        if not params.project_name and params.file_path:
+            # Use parent directory name as project name
+            path = Path(params.file_path)
+            if path.parent.name:
+                params.project_name = path.parent.name
+
+        return params
+
+    def to_dict(self) -> dict[str, Any]:
+        """Convert to dictionary for query execution.
+
+        Returns:
+            Dictionary of non-None parameters
+        """
+        return {key: value for key, value in self.__dict__.items() if value is not None}
+
+
+class ContextQueryExecutor:
+    """Executes context queries for agents."""
+
+    def __init__(self, settings: ScriptRAGSettings | None = None):
+        """Initialize the executor.
+
+        Args:
+            settings: Configuration settings
+        """
+        if settings is None:
+            from scriptrag.config import get_settings
+
+            settings = get_settings()
+
+        self.settings = settings
+        self.engine = QueryEngine(settings)
+
+    async def execute(
+        self,
+        query_sql: str,
+        parameters: ContextParameters,
+    ) -> list[dict[str, Any]]:
+        """Execute a context query with parameters.
+
+        Args:
+            query_sql: SQL query string
+            parameters: Context parameters
+
+        Returns:
+            List of result rows as dictionaries
+        """
+        if not query_sql:
+            return []
+
+        try:
+            # Convert SQL to QuerySpec for validation
+            spec = self._sql_to_spec(query_sql, parameters)
+
+            # Execute using QueryEngine
+            params_dict = parameters.to_dict()
+            rows, execution_time = self.engine.execute(spec, params_dict)
+
+            logger.info(
+                "Context query executed successfully",
+                row_count=len(rows),
+                execution_time_ms=execution_time,
+                params=list(params_dict.keys()),
+            )
+
+            return rows
+
+        except Exception as e:
+            logger.error(
+                "Context query execution failed",
+                error=str(e),
+                query_preview=query_sql[:200] if query_sql else None,
+            )
+            # Return empty results on failure (graceful degradation)
+            return []
+
+    def _sql_to_spec(self, sql: str, parameters: ContextParameters) -> QuerySpec:
+        """Convert raw SQL to QuerySpec.
+
+        Args:
+            sql: SQL query string
+            parameters: Context parameters for validation
+
+        Returns:
+            QuerySpec instance
+        """
+        # Extract parameter names from SQL
```

Comment:
If we would pass the default parameters as a typed object we would not need most of this. Given that the parameters are not user-extensible, this would be the right choice.